OpenDreamShared/Compiler/DM/DMParser.cs

```diff
@@ -418,13 +418,56 @@ public class DMParser : Parser<Token> {
                 Whitespace();
                 DMASTPath path = Path();
                 if (path == null) Error("Expected a variable name");
-                Whitespace();
 
                 DMASTExpression value = null;
 
-                if (Check(TokenType.DM_Equals)) {
+                if (Check(TokenType.DM_LeftBracket)) //TODO: Multidimensional lists
+                {
+                    //Type information
+                    if (path.Path.FindElement("list") != 0)
+                    {
+                        path = new DMASTPath(new DreamPath("list/" + path.Path.PathString));
+                    }
+
                     Whitespace();
-                    value = Expression();
+
+                    DMASTConstantInteger size = null;
+                    if (!Check(TokenType.DM_RightBracket))
+                    {
+                        var expr = Expression() as DMASTExpressionConstant;
+                        if (expr is DMASTConstantInteger integer)
+                        {
+                            size = integer;
+                        }
+                        else
+                        {
+                            size = new DMASTConstantInteger(0);
+                        }
+
+                        Whitespace();
+                        Consume(TokenType.DM_RightBracket, "Expected ']'");
+                    }
```

Comment:
`Expression()` returns null if it couldn't parse an expression, and already consumes any whitespace after it. `var/L[expression]` is also always equal to `var/list/L = new /list(expression)` (from what I can tell). Because of this, this can be simplified down to 2 lines:

```suggestion
                    DMASTExpression size = Expression();
                    Consume(TokenType.DM_RightBracket, "Expected ']'");
```